include/nil/crypto3/math/multithreading/thread_pool.hpp

```diff
@@ -0,0 +1,117 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+//---------------------------------------------------------------------------//
+
+#ifndef CRYPTO3_THREAD_POOL_HPP
+#define CRYPTO3_THREAD_POOL_HPP
+
+#include <boost/asio/thread_pool.hpp>
+#include <boost/asio/post.hpp>
+
+#include <functional>
+#include <future>
+#include <thread>
+#include <limits>
+#include <memory>
+#include <stdexcept>
+
+
+namespace nil {
+    namespace crypto3 {
+
+        class ThreadPool {
+        public:
+
+            static ThreadPool& get_instance(std::size_t pool_number, std::size_t pool_size = std::thread::hardware_concurrency()) {
+                static ThreadPool instance0(0, pool_size);
+                static ThreadPool instance1(1, pool_size);
+                
+                if (pool_number == 0)
+                    return instance0;
+                if (pool_number == 1)
+                    return instance1;
+                throw std::invalid_argument("Invalid instance of thread pool requested.");
+            }
+
+            ThreadPool(const ThreadPool& obj)= delete;
+            ThreadPool& operator=(const ThreadPool& obj)= delete;
+
+            template<class ReturnType>
+            inline std::future<ReturnType> post(std::function<ReturnType()> task) {
+                auto packaged_task = std::make_shared<std::packaged_task<ReturnType()>>(std::move(task));
+                std::future<ReturnType> fut = packaged_task->get_future();
+                boost::asio::post(pool, [packaged_task]() -> void { (*packaged_task)(); });
+                return fut;
+            }
+ 
+            // Waits for all the tasks to complete.
+            inline void join() {
+                pool.join();
+            }
+
+            // Divides work into a ranges and makes calls to func in parallel.
+            template<class ReturnType>
+            std::vector<std::future<ReturnType>> block_execution(
+                    std::size_t elements_count,
+                    std::function<ReturnType(std::size_t begin, std::size_t end)> func) {
+
+                std::vector<std::future<ReturnType>> fut;
+                std::size_t cpu_usage = std::max((size_t)1, std::min(elements_count, pool_size));
+                std::size_t element_per_cpu = elements_count / cpu_usage;
+
+                // Pool #0 will take care of the lowest level of operations, like polynomial operations.
+                // We want the minimal size of element_per_cpu to be 65536, otherwise the cores are not loaded.
+                if (pool_number == 0 && element_per_cpu < 65536) {
```

Comment:
Since pool numbers bear special significance, I suggest using an enum instead of the magic constants 0 and 1.

Reply:
Done.